### ISSUE TITLE
Hint to catch redundant //ignores

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#35234]
It'd be great if there was a lint that would trigger whenever there was an `// ignore` that wasn't actively silencing an analyzer message. This would let us catch old ignores that work around bugs that have since been fixed, or ignores that we don't need because we disabled the relevant lint, etc.

cc @a14n

**Comment by bwilkerson** [comment]
I completely agree that we need this. It might even be reasonable to make it be opt-in.

However, we can't implement it as a normal lint because there's no way to guarantee that all of the other lints will have already been run. This would need to run after all other diagnostics have been generated, otherwise we won't be able to determine whether a given ignore is valid.

**Comment by bwilkerson** [comment]
Because it can't be implemented as a lint, the linter issue has been closed.

**Comment by srawlins** [comment]
I propose the following static analysis rule:

A new error code (a HintCode) called something like `unused_ignore_comment`.

This error code cannot be ignored with an `ignore:` or `ignore_for_file:` comment. It can be ignored via `analysis_options.yaml`.

Only unused ignore comments for error codes _which exist_ are reported. For example, a typo like `// ignore: daed_code` does not ignore any real error, but no `unused_ignore_comment` Hint will be reported, since there exists no error code, `daed_code`.

This new error code is not intended to catch typos. It is intended to catch ignore comments which once were used, and are no longer.

If we did report comments like `// ignore: to disregard` and `// Some things I wish I could\n// ignore: dead code for example`, I think users would be fighting a lot of unintentional "ignore comments." Which, as stated above, themselves cannot be ignored from within the source file.

**Comment by bwilkerson** [comment]
> This error code cannot be ignored with an `ignore:` or `ignore_for_file:` comment.

I don't see any value in adding this special case handling. I think the number of people who will choose to ignore this hint will be very small and I don't see why we'd disallow it. Besides, if we want a way to make certain diagnostics impossible to ignore then that mechanism would cover this case without special casing.

> ... I think users would be fighting a lot of unintentional "ignore comments."

I doubt that's true, but we could gather evidence easily enough.

As a counter argument, though, I'll note that we recently removed a diagnostic (because the error was removed from the spec). If there are any ignore comments for that diagnostic, your proposed limitation would prevent this lint from helping users clean up their code. I think we could safely not generate diagnostics for comments that are not of the right form, but I'd prefer for the hint to work if what follows the `ignore:` is a comma separated list of identifiers. That should cover the examples you gave while still allowing it to catch errors that have been removed.

**Comment by srawlins** [comment]
> I don't see any value in adding this special case handling. I think the number of people who will choose to ignore this hint will be very small and I don't see why we'd disallow it.

I thought it would be a lot of effort to allow ignoring these, but I think I was wrong. These should be ignorable.

**Comment by srawlins** [comment]
I see the flutter repo using the ignore syntax like this in a dozen places:

https://github.com/flutter/flutter/blob/b9a56b9f48462d897101284f5f57b8568ef683c6/packages/flutter/lib/src/cupertino/text_theme.dart#L122

```
// ignore: avoid_unused_constructor_parameters, the parameter is deprecated.
```

I guess at a minimum I would ignore "ignored codes" with whitespace (don't call them out as unused). Also I've seen this:

```
// ignore: prefer_void_to_null, https://github.com/dart-lang/sdk/issues/34416
```

So... ignore "ignored codes" with... a list of punctuation characters? colons, slashes?

**Comment by bwilkerson** [comment]
The way analyzer processes ignore comments is to take everything following `ignore:`, split it on commas, trim off white space at the beginning and end, and convert to lower case. That gives it a list of codes to ignore. This hint should process them the same way, and should share as much code as possible with `IgnoreInfo` in order to ensure consistency should that we change the way the comments are processed.

Every valid code name is a valid identifier. In both of the examples you gave we'd end up with items in the list that are not valid identifiers. I suspect we'll get the best results by discarding / ignoring anything that isn't a valid identifier, and then validating that each remaining identifier is a known code and that a diagnostic with that code is actually ignored at that location.

That hint processing will need to happen after all of the diagnostics have been computed, including lints, in order to know whether a diagnostic is reported at that location. Note that that means that it might be harder than I'd first assumed to make this hint be ignorable. If that's the case, then I don't think we should do the extra work until we know of a valid use case for allowing them to be ignored.

> A new error code (a HintCode) called something like `unused_ignore_comment`.

I'm not sold on the name because of the fact that a single comment can contain both valid (used) and invalid (unused) codes. In that case the comment isn't unused, just one of the codes. Maybe "code" instead of "comment"?

**Comment by srawlins** [comment]
I've got a WIP if anyone is interested. https://dart-review.googlesource.com/c/sdk/+/126620

**Comment by srawlins** [comment]
@Hixie how will Flutter deal with "unused" ignore comments and Flutter's `dev/analyze/bots.dart`'s `verifyDeprecations`?

https://github.com/flutter/flutter/blob/b9a56b9f48462d897101284f5f57b8568ef683c6/dev/bots/analyze.dart#L167

I have a PR to remove "unused" ignore comments in flutter/flutter, but the PR breaks on instances of `flutter_deprecation_syntax`: https://github.com/flutter/flutter/pull/45855/checks?check_run_id=327666716

**Comment by srawlins** [comment]
Same question for [`flutter_tools/test/general.shard/forbidden_imports_test.dart`](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/forbidden_imports_test.dart), which looks at source files for certain patterns, allowable via, e.g. `ignore: dart_convert_import`, `ignore: package_path_import`.

Maybe just a new way to ignore these things? `// flutter_ignore:`? `// ignore_for_flutter:`?

**Comment by Hixie** [comment]
feel free to change the syntax that analyze.dart uses, it doesn't need to be `//ignore`. I just used that because it fit the general pattern. Ideally something prettier than "ignore_for_flutter"...

e.g. `// this deprecation notice is grandfathered-in for historical reasons` or something

**Comment by bwilkerson** [comment]
As of https://dart-review.googlesource.com/c/sdk/+/156402 we now have the framework in place to catch these issues. We are only reporting on duplicated //ignore comments (where the same diagnostic is being ignored more than once) because the rest can't be enabled until some changes are made in the Flutter code base.

**Comment by Hixie** [comment]
What do I enable to see the lint?

**Comment by bwilkerson** [comment]
It's enabled by default (so it's technically a hint, though the distinction is more distracting than meaningful). I don't know whether that CL has rolled into the Flutter SDK at this point, but when it does it ought to catch cases like
```
void f() {
  // ignore: unused_local_variable, unused_local_variable
  var x = 0;
}
```
and
```
// ignore_for_file: unused_local_variable
void f() {
  // ignore: unused_local_variable
  var x = 0;
}
```
There weren't any cases of this in Flutter, which is why I could enable that particular check.

The checks for ignores of errors that can't be ignored, and of ignores of codes that aren't being generated, both are waiting for changes in the Flutter code base so they're currently disabled.

**Comment by Hixie** [comment]
Sorry, I meant, how do I enable the lint that can't be enabled by default in Flutter yet?

**Comment by bwilkerson** [comment]
I uploaded a CL with the necessary changes (https://dart-review.googlesource.com/c/sdk/+/168561), and you can see the Flutter failures here: https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8865911342565319552/+/steps/analyze_flutter_flutter/0/stdout.

**Comment by a14n** [comment]
sometimes we have comment that give some context:

```
// ignore: avoid_equals_and_hash_code_on_mutable_classes, to remove in NNBD with a late final hashcode
```

In this case for instance there's the lint:

> info • The diagnostic 'to remove in nnbd with a late final hashcode' isn't produced at this location so it doesn't need to be ignored • packages/flutter/lib/src/widgets/shortcuts.dart:102:61 • unnecessary_ignore

What pattern could we use to keep the comment?

**Comment by bwilkerson** [comment]
One possibility would be to put the comment on a separate line above the ignore comment.

**Comment by a14n** [comment]
CL for flutter to remove unused ignore : flutter/flutter#68629

**Comment by Hixie** [comment]
@bwilkerson Is there any chance we can change the way this lint works? This syntax:

```dart
  foo() // ignore: bar, explanation for ignore
```

...is aesthetically more pleasing, and IMHO clearer, than this:

```dart
  // explanation for ignore
  // ignore: bar
  foo();
```

...or:

```dart
  // explanation for ignore
  foo(); // ignore: bar
```

...because when you're reading the code the explanation makes no sense until you see the ignore it's referencing, and the ignore makes no sense until you see the code that is triggering the false-positive in the analyzer.

**Comment by bwilkerson** [comment]
> Is there any chance we can change the way this lint works?

Yes, of course.

The ignore syntax currently allows either `ignore:` or `ignore_for_file:` followed by a comma separated list of diagnostic names. I suspect that the problem is that `explanation for ignore` is being treated as a diagnostic name and therefore being found to be invalid.

One possible solution would be to ignore any "name" that isn't a valid identifier. That's probably reasonable because you probably don't have any single word explanations. Does that sound like what you're looking for?

**Comment by Hixie** [comment]
catching `// ignore: invalid_lint_name` seems valuable
I wasn't aware you could list multiple lints on a line. Is that common? I don't think I've ever seen it... It seems less critical a use case than documenting why a particular lint is present at all.

**Comment by bwilkerson** [comment]
I don't think I've ever done it, but it's been supported for a long time and I don't know whether it would break user code to remove the support.

**Comment by srawlins** [comment]
> Is that common?

It is for `ignore_for_file:` in code generators, I think.

> I don't know whether it would break user code to remove the support.

I think it's necessary, the way they attach to specific lines now.

```dart
// ignore: a
// ignore: b
code;
```

would _not_ suppress a diagnostic, `a` reported on the line containing `code;`.

We could also change how they attach; you could imagine forgiving heuristics.

**Comment by srawlins** [comment]
> One possible solution would be to ignore any "name" that isn't a valid identifier.

+1

**Comment by Hixie** [comment]
FWIW, I'd be fine with removing "ignore_for_file" entirely, it seems like a back door to technical debt to me. We frown on it in the Flutter code base (it counts much more towards our tech debt benchmark). Also I prefer // ignore at the end of lines than before lines, where you could do `//ignore: a // ignore: b` if you really needed to. Plus I think people shouldn't be ignoring lints in generated code either, that's another source of technical debt.

anyway, how about this syntax, it seems like it would satisfy all our needs:

`// ignore: code (explanation), code (explanation)`

...where the parenthetical is optional. Lets us do multi-lint ignores, attach before and after, allows for documentation of why the lint should be ignored, allows for catching invalid lints, and avoids the need for heuristics, which seems to be all the features that were requested here.

**Comment by bwilkerson** [comment]
> I'd be fine with removing "ignore_for_file" entirely ...

I actually think there are a few places where it's useful and not harmful, but personal opinion aside it would be a breaking change to remove it and I think there are too many other critical issues right now to add another one to the list.

> ... how about this syntax ...

I like it. It's a nice incremental improvement that solves at least one problem.

**Comment by Hixie** [comment]
I'd love for that syntax to be available. I've struggled with trying to document why lints exist in code recently because it literally requires three lines per lint which is a lot of space to be giving something you're trying to ignore.

**Comment by goderbauer** [comment]
I've cleaned up all the occurrences of `// ignore:` in flutter's code base that weren't ignoring anything. However, we still have cases of

```
// ignore: foo_bar, explanation bla bla
```

in the code base.

@bwilkerson We still need to wait for that syntax discussed above to be available before we can turn on https://dart-review.googlesource.com/c/sdk/+/168561, right?

**Comment by goderbauer** [comment]
Looks like we also need to change our `// ignore: dart_io_import` syntax (and its friends) to something that doesn't clash with this. Maybe something like `// flutter_ignore: dart_io_import`?

@bwilkerson How can I run https://dart-review.googlesource.com/c/sdk/+/168561 over the latest flutter code base to see the places that need updating?

**Comment by bwilkerson** [comment]
I rebased the CL and started the try bots, including the flutter-analyze bot. We should see what's failing when they're finished.

**Comment by goderbauer** [comment]
> I rebased the CL and started the try bots, including the flutter-analyze bot. We should see what's failing when they're finished.

Looks like my clean-up PRs didn't actually get submitted due to some problems in google3. I'll look into that and will ping you once it is resolved to restart the bot... Sorry about that.

**Comment by bwilkerson** [comment]
No problem. It's easy to click a button. :-)

**Comment by goderbauer** [comment]
The check is complaining that this ignore is not needed, but if I remove it, the analyzer is not happy: https://github.com/flutter/flutter/blob/c0bcb4fc16381ea0d9e2ff8b22843ab961167308/packages/flutter/test/widgets/shortcuts_test.dart#L71

Same for the other two `// ignore: prefer_const_literals_to_create_immutables` in that file.

**Comment by goderbauer** [comment]
@bwilkerson Can you kick the bot one more time?

**Comment by bwilkerson** [comment]
Done

**Comment by goderbauer** [comment]
Looks like the engine is completely clean now. There are a handful of ignores I missed in the framework, those are getting cleaned up here: https://github.com/flutter/flutter/pull/79071.

In previous runs of the bot it would complain about explanations we had added to an ignore, like:

```
   info • The diagnostic 'this is the original sin' isn't produced at this location so it doesn't need to be ignored • packages/flutter/lib/src/widgets/framework.dart:760:61 • unnecessary_ignore
   info • The diagnostic 'https://github.com/dart-lang/linter/issues/2345' isn't produced at this location so it doesn't need to be ignored • packages/flutter/lib/src/widgets/implicit_animations.dart:302:111 • unnecessary_ignore
```

Newer runs don't complain about that anymore even though its still in our codebase:

https://github.com/flutter/flutter/blob/bed88a1a5e0944cf9134d19119abc508b8fd2b7b/packages/flutter/lib/src/widgets/implicit_animations.dart#L302
https://github.com/flutter/flutter/blob/bed88a1a5e0944cf9134d19119abc508b8fd2b7b/packages/flutter/lib/src/widgets/framework.dart#L760

Is that an expected improvement of this hint?

**Comment by goderbauer** [comment]
Looks like what I pointed out in my previous comment is due to improvements made in https://dart-review.googlesource.com/c/sdk/+/192725. All good then!

**Comment by goderbauer** [comment]
Some `// ignore`s in generated code are still problematic, details in https://github.com/flutter/flutter/issues/79091.

**Comment by pq** [comment]
As per a conversation w/ @bwilkerson, I'm picking this up again.

Because this would be so impacting as a warning, I plan to approach this as a new lint: `unnecessary_ignore`.

For interested folks, I'd be curious how you feel about generated files. My inclination is to disable reporting on generated files but I'd be curious if there are any counter arguments?

**Comment by rrousselGit** [comment]
I don't mind either way for generated files. Generated files can generate an extra `ignore_for_file: unnecessary_ignore` if need be.


What I'm mainly worried about is analyzer plug-ins, and SDK versioning.

Say we have `// ignore: my_custom_lint`.
Would unnecessary_ignore complain? Because that lint could be from an analyzer plug-in or a newer version of the SDK.

I both use custom lint rules and check my packages against multiple Dart versions. Meaning my codebase may have ignores that Dart doesn't recognize

**Comment by pq** [comment]
> What I'm mainly worried about is analyzer plug-ins, and SDK versioning.

Excellent question.

@srawlins: I'm assuming that plugins get their own lint registry that's run separately from the "core lints" (in their own isolate). If that's right we'll have to (minimally) make sure we don't over-report... Thoughts?

**Comment by srawlins** [comment]
> I'm assuming that plugins get their own lint registry that's run separately from the "core lints" (in their own isolate). If that's right we'll have to (minimally) make sure we don't over-report... Thoughts?

Correct. The current plan is that ignoring plugin diagnostics inline (`// ignore:` and `// ignore_for_file:`) will require namespacing. Something like `// ignore: plugin_1.diagnostic_name`.

So the unnecessary ignore rule should probably ignore... anything that's not a simple identifier?

As for SDK versions, the only solution I've thought of is more ignores :P

```dart
// Some people are using Dart 4.3, where we have the "dont_use_seven" rule.
// And some are using Dart 4.2. So we have to ignore unnecessary_ignore, in case
// the active SDK is not aware of the "dont_use_seven" rule.
// ignore: unnecessary_ignore
// ignore: dont_use_seven
int x = 7;
```

Don't ask me what this looks like if you also have the [document_ignores](https://dart.dev/tools/linter-rules/document_ignores) rule enabled.

**Comment by rrousselGit** [comment]
> Correct. The current plan is that ignoring plugin diagnostics inline (`// ignore:` and `// ignore_for_file:`) will require namespacing. Something like `// ignore: plugin_1.diagnostic_name`.
> 
> So the unnecessary ignore rule should probably ignore... anything that's not a simple identifier?

Depending on how new plugins work, they could be required to list all Error Codes.

Then, analyzer could know if `ignore: code` matches something from a plugin or whether it's just useless given the current environment

**Comment by rrousselGit** [comment]
> As for SDK versions, the only solution I've thought of is more ignores :P
> 
> ```dart
> // Some people are using Dart 4.3, where we have the "dont_use_seven" rule.
> // And some are using Dart 4.2. So we have to ignore unnecessary_ignore, in case
> // the active SDK is not aware of the "dont_use_seven" rule.
> // ignore: unnecessary_ignore
> // ignore: dont_use_seven
> int x = 7;
> ```

I wonder if we could use some form of version constraint in the ignore to have them SDK/package aware

Like:

```dart
// ignore: code @SDK=^3.1.0
<some code that warns after 3.1.0
```

Although that's probably a lot for not much.

**Comment by pq** [comment]
> Depending on how new plugins work, they could be required to list all Error Codes.

Or maybe the plugin isolate could communicate a list of registered lints *back* to the main analyzer isolate when lints are added to the plugin's lint registry?

**Comment by pq** [comment]
Thanks again for all the feedback.  A basic implementation of an `unnecessary_ignore` lint has been landed w/ associated fixes.

Feel free to open new issues w/ any issues (and cc me).

Thanks!
